framework/web/Request.php

```diff
@@ -1131,15 +1142,53 @@ public function getUserAgent()
      */
     public function getUserIP()
     {
-        foreach ($this->ipHeaders as $ipHeader) {
+        foreach($this->getTrustedIpHeaders() as $ipHeader) {
             if ($this->headers->has($ipHeader)) {
-                return trim(explode(',', $this->headers->get($ipHeader))[0]);
+                $ip = $this->getUserIpFromIpHeader($this->headers->get($ipHeader));
+                if ($ip !== null) {
+                    return $ip;
+                }
             }
         }
 
         return $this->getRemoteIP();
     }
 
+    /**
+     * Return user IP's from IP header.
+     * @param string $ips comma separated IP list
+     * @return string|null
+     * @see $getUserHost
+     * @see $ipHeader
+     * @see $trustedHeaders
+     * @since 2.0.28
+     */
+    protected function getUserIpFromIpHeader($ips)
+    {
+        $ips = preg_split('/\s*,\s*/', trim($ips), -1, PREG_SPLIT_NO_EMPTY);
+        krsort($ips);
+        $validator = $this->getIpValidator();
+        $resultIp = null;
+        foreach ($ips as $ip) {
+            $stop = true;
+            foreach ($this->trustedHosts as $cidr => $cidrOrHeaders) {
+                if (!is_array($cidrOrHeaders)) {
+                    $cidr = $cidrOrHeaders;
+                }
+                $validator->setRanges($cidr);
+                $resultIp = $ip;
+                if ($validator->validate($ip)) {
+                    $stop = false;
+                    break;
+                }
+            }
+            if ($stop) {
+                break;
+            }
+        }
+        return $resultIp;
```

Comment:
This looks suspicious. `$resultIp` is set before `$validator->validate($ip)`, so at this point it could contain invalid IP?

Reply:
validation added

Reply:
Why `$resultIp = $ip;` is outside of `$validator->validate($ip)` condition?

Reply:
I moved it to the main cycle. 

> Why $resultIp = $ip; is outside of $validator->validate($ip) condition?

After the last trusted IP address is the usable IP address. The result may be invalid IP or even nonexistent. Therefore, at the very end of the method, it validates as a plus.